atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraClientPool.java

```diff
@@ -264,13 +267,24 @@ private boolean isHostHealthy(InetSocketAddress host) {
     }
 
     private CassandraClientPoolingContainer getRandomGoodHost() {
+        return getRandomGoodHostForPredicate(address -> true);
+    }
+
+    @VisibleForTesting
+    CassandraClientPoolingContainer getRandomGoodHostForPredicate(Predicate<InetSocketAddress> predicate) {
         Map<InetSocketAddress, CassandraClientPoolingContainer> pools = currentPools;
 
-        Set<InetSocketAddress> livingHosts = Sets.difference(pools.keySet(), blacklistedHosts.keySet());
+        Set<InetSocketAddress> filteredHosts = pools.keySet().stream().filter(predicate).collect(Collectors.toSet());
+        if (filteredHosts.isEmpty()) {
+            log.error("No hosts match the provided predicate. We will ignore said predicate.");
```

Comment:
This seems super weird. We should throw or return no container and do this logic one level up